### PR TITLE
Go Mod Validator Fixes

### PR DESCRIPTION
**Pull request by HenryNguyen5**
Addresses:
- https://smartcontract-it.atlassian.net/browse/RE-2924
- https://smartcontract-it.atlassian.net/browse/RE-2889
- https://smartcontract-it.atlassian.net/browse/RE-2925

### CODE REVIEW

File: apps/go-mod-validator/src/github.ts [hunk 1]
```diff
@@ -24,104 +27,95 @@ async function isCommitInDefaultBranch(
 }
 
 async function isTagInDefaultBranch(
```
Comment (erikburt) [inline apps/go-mod-validator/src/github.ts:29]: This still doesn't deal with annotated tags right?

File: apps/go-mod-validator/src/go-mod-validator.ts [hunk 2]
```diff
@@ -43,13 +43,29 @@ export async function run(): Promise<string> {
...
-  const errs: Map<GoModule, string> = new Map();
-
+  const errs: Map<BaseGoModule, string> = new Map();
   const validating = depsToValidate.map(async (d) => {
-    const isValid = await isGoModReferencingDefaultBranch(d, gh);
+    // Bit of a code smell, but I wanted to avoid adding the defaultBranchGetter to deps.ts to keep it separate from
+    // the GitHub API client.
+    // And we want the default branch available in this scope for context.
+    const defaultBranch = await getDefaultBranch(gh, d);
+    const isValid = await isGoModReferencingDefaultBranch(d, defaultBranch, gh);
```
Comment (erikburt) [inline apps/go-mod-validator/src/go-mod-validator.ts:52]: nit: the ordering of the params is inconsistent

File: apps/go-mod-validator/src/go-mod-validator.ts [hunk 3]
```diff
@@ -43,13 +43,29 @@ export async function run(): Promise<string> {
...
+      parsedVersion = d.tag;
+    }
 
     if (!isValid) {
-      errs.set(d, "dependency not on default branch");
+      errs.set(
+        d,
+        `[${d.goModFilePath}] dependency ${d.name} not on default branch.
+Default branch: ${defaultBranch}
+Version: ${parsedVersion}`,
```
Comment (erikburt) [inline apps/go-mod-validator/src/go-mod-validator.ts:67]: Does `d.tag ?? d.commitSha ?? "UNKNOWN"` work here?

File: apps/go-mod-validator/src/pseudo-version.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  const { rev } = parsePseudoVersion(v);
+  return rev;
+}
+
+function parsePseudoVersion(v: string): {
+  base: string;
+  timestamp: string;
+  rev: string;
+  build: string;
+  err: Error | null;
```
Comment (erikburt) [inline apps/go-mod-validator/src/pseudo-version.ts:53]: `err` is always null

File: apps/go-mod-validator/test/__snapshots__/deps.test.ts.snap [hunk 5]
```diff
@@ -0,0 +1,99 @@
+// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html
```
Comment (erikburt) [inline apps/go-mod-validator/test/__snapshots__/deps.test.ts.snap:1]: This seems like a weird thing to snapshot. The result of the method should never change and this leaves it open to accidentally updating the snapshot.

**Review by erikburt — CHANGES_REQUESTED** [review]
Missing a changeset as well.

File: apps/go-mod-validator/src/deps.ts [hunk 6]
```diff
@@ -172,15 +192,17 @@ export async function getDeps(
         encoding: "utf-8",
         cwd: dir,
       });
-      core.debug(`Raw output: ${output}`);
+      process.env.CI && core.debug(`Raw output: ${output}`);
```
[the remark this answers is not in the document]
Comment (HenryNguyen5) [reply apps/go-mod-validator/src/deps.ts:195]: `CI` is always present as an env var during runs inside GHA

Makes sense to mock it out

File: apps/go-mod-validator/src/github.ts (hunk 1 above)
Comment (HenryNguyen5) [reply apps/go-mod-validator/src/github.ts:29]: It does, here's the output from the api call for `v2.15.0` from `chainlink`
``` json
{
    "name": "v2.15.0",
    "zipball_url": "https://api.github.com/repos/smartcontractkit/chainlink/zipball/refs/tags/v2.15.0",
    "tarball_url": "https://api.github.com/repos/smartcontractkit/chainlink/tarball/refs/tags/v2.15.0",
    "commit": {
      "sha": "883c9f9059c259ca17454703b57521ea9ac166cb",
      "url": "https://api.github.com/repos/smartcontractkit/chainlink/commits/883c9f9059c259ca17454703b57521ea9ac166cb"
    },
    "node_id": "MDM6UmVmMTExNDU1ODY3OnJlZnMvdGFncy92Mi4xNS4w"
  }
```
The tag's object ref is different
```
5f9c6900a76372940347436a4a0dbd713b710323 refs/tags/v2.15.0
```

File: apps/go-mod-validator/src/go-mod-validator.ts (hunk 3 above)
Comment (HenryNguyen5) [reply apps/go-mod-validator/src/go-mod-validator.ts:67]: Unfortunately not, the typechecker is not smart enough to realize that im still handling the unions correctly

**Review by erikburt — APPROVED** [review]

**Pull request merged**